plugins/task-witness/controller/task_witness_deploy.py

```
@@ -9449,6 +9487,90 @@ def _validate_retained_control_preimage(
+    sequences = [item[0]["sequence"] for item in chain]
+    expected_sequences = list(range(1, len(chain) + 1))
+    by_sequence = {value["sequence"]: (value, raw) for value, raw in chain}
+    if sequences != expected_sequences or len(by_sequence) != len(chain):
+        raise DeploymentError("B1 retained deployment chain is not exact")
+    if initial_receipt_profile == BRIDGE_LEGACY_RECEIPT_PROFILE:
+        if (
+            bridge_transition_seen
+            or sequences != [1, 2]
+            or by_sequence[2] != (initial_receipt, initial_receipt_raw)
+        ):
+            raise DeploymentError("B1 retained deployment chain is not exact")
```

Ruff (0.16.3)

[warning] Avoid specifying long messages outside the exception class

(TRY003)